### PR TITLE
Allow overriding of values from StandaloneConfig by repo-config.csv #390

### LINKED ISSUE #390: Revamp the override option
As discussed earlier with Akshay, there doesn't seems to be a clean way to override just a specific field for the json config.

Some ideas that we can possibly revamp this:
1. Make the cli param highest order which can override any single field
2. Make single field override-able option, to override some fields of the json config (if any), in the csv config instead of override the entire thing.

Can discuss other alternatives.

### PR DESCRIPTION
**Pull request by myteo**
Fixes #390.

```
There is no way to override a specific field of the Standalone config.

It is not possible for users to use settings from Standalone config for
certain fields and settings from CSV config file for other fields.

Let's override a specific field of the Standalone config if a field
value in repo-config.csv is prefixed with "override:".
```

### CODE REVIEW

**Comment by myteo** [comment]
@reposense/devs Ready for review.

File: src/main/java/reposense/parser/RepoConfigCsvParser.java [hunk 1]
```diff
@@ -46,9 +46,22 @@ public RepoConfigCsvParser(Path csvFilePath) throws IOException {
     protected void processLine(List<RepoConfiguration> results, String[] elements) throws InvalidLocationException {
         RepoLocation location = new RepoLocation(getValueInElement(elements, LOCATION_POSITION));
         String branch = getValueInElement(elements, BRANCH_POSITION, RepoConfiguration.DEFAULT_BRANCH);
+        boolean isFormatsOverriding = isElementOverridingStandaloneConfig(elements, FILE_FORMATS_POSITION);
+        if (isFormatsOverriding) {
+            removeOverrideKeywordFromElement(elements, FILE_FORMATS_POSITION);
+        }
         List<Format> formats = Format.convertStringsToFormats(getManyValueInElement(elements, FILE_FORMATS_POSITION));
+        boolean isIgnoreGlobListOverriding = isElementOverridingStandaloneConfig(elements, IGNORE_GLOB_LIST_POSITION);
```
Comment (yamidark) [inline src/main/java/reposense/parser/RepoConfigCsvParser.java]: Let's leave a whitespace between the getting of each value here, make it look neater

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 2]
```diff
@@ -247,4 +249,23 @@ public void repoConfig_defaultBranch_success() throws ParseException, IOExceptio
...
+    @Test
+    public void repoConfig_overrideKeyword_success() throws ParseException, IOException, HelpScreenException {
+        RepoConfigCsvParser repoConfigCsvParser = new RepoConfigCsvParser(REPO_CONFIG_OVERRIDE_KEYWORD_FILE);
+        List<RepoConfiguration> configs = repoConfigCsvParser.parse();
+        RepoConfiguration config = configs.get(0);
+
+        Assert.assertEquals(1, configs.size());
+        Assert.assertEquals(new RepoLocation(TEST_REPO_BETA_LOCATION), config.getLocation());
+        Assert.assertEquals(TEST_REPO_BETA_BRANCH, config.getBranch());
+        Assert.assertEquals(TEST_REPO_BETA_CONFIG_FORMATS, TEST_REPO_BETA_CONFIG_FORMATS);
```
Comment (yamidark) [inline src/test/java/reposense/parser/CsvParserTest.java]: ??? Testing the same 2 values?

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 3]
```diff
@@ -247,4 +249,23 @@ public void repoConfig_defaultBranch_success() throws ParseException, IOExceptio
...
+        Assert.assertFalse(config.isStandaloneConfigIgnored());
+        Assert.assertEquals(config.getIgnoreCommitList(),
+                CommitHash.convertStringsToCommits(TEST_REPO_BETA_CONFIG_IGNORED_COMMITS));
```
Comment (yamidark) [inline src/test/java/reposense/parser/CsvParserTest.java]: By convention, when asserting 2 elements are equal, 'expected' should be the 1st argument, and the actual should be the 2nd.

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 4]
```diff
@@ -247,4 +249,23 @@ public void repoConfig_defaultBranch_success() throws ParseException, IOExceptio
         Assert.assertEquals(expectedConfig.getBranch(), actualConfigs.get(0).getBranch());
         Assert.assertEquals(expectedConfig.getBranch(), authorConfigs.get(0).getBranch());
     }
+
+    @Test
+    public void repoConfig_overrideKeyword_success() throws ParseException, IOException, HelpScreenException {
```
Comment (yamidark) [inline src/test/java/reposense/parser/CsvParserTest.java]: Do add on to the other `repoConfig` test, to check that the 'override' value for the elements are `false`

File: src/test/java/reposense/parser/RepoConfigurationTest.java [hunk 5]
```diff
@@ -42,6 +42,8 @@
             .getResource("RepoConfigurationTest/repoconfig_formats_test").getFile()).toPath();
     private static final Path WITHOUT_FORMATS_TEST_CONFIG_FILES = new File(CsvParserTest.class.getClassLoader()
             .getResource("RepoConfigurationTest/repoconfig_withoutformats_test").getFile()).toPath();
+    private static final Path OVERRIDE_STANDALONE_TEST = new File(CsvParserTest.class.getClassLoader()
```
Comment (yamidark) [inline src/test/java/reposense/parser/RepoConfigurationTest.java]: Append `_CONFIG_FILE` to the back

File: src/test/java/reposense/parser/RepoConfigurationTest.java [hunk 6]
```diff
@@ -260,4 +262,39 @@ public void repoConfig_sameLocationDifferentBranch_notEqual() throws InvalidLoca
...
+            throws ParseException, GitCloneException, IOException, HelpScreenException {
+        RepoConfiguration expectedConfig = new RepoConfiguration(new RepoLocation(TEST_REPO_DELTA), "master");
+
+        List<Author> expectedAuthorList = new ArrayList<>();
+        Author[] authors = new Author[]{FIRST_AUTHOR, SECOND_AUTHOR, THIRD_AUTHOR, FOURTH_AUTHOR};
+        for (Author author : authors) {
+            Author expectedAuthor = new Author(author);
+            List<String> expectedAuthorIgnoreGlobList = new ArrayList<>();
+            expectedAuthorIgnoreGlobList.addAll(author.getIgnoreGlobList());
+            expectedAuthorIgnoreGlobList.removeAll(REPO_LEVEL_GLOB_LIST);
```
Comment (yamidark) [inline src/test/java/reposense/parser/RepoConfigurationTest.java]: Add a comment here, explaning that the `Author`s original `ignoreGlobList` contains values from standalone config repo level, thus needs to be removed here

File: src/test/java/reposense/parser/CsvParserTest.java (hunk 2 above)
Comment (myteo) [reply src/test/java/reposense/parser/CsvParserTest.java]: Oops.. Blindly copy and pasted this part from the existing test😅 Will fix that one as well.

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 7]
```diff
@@ -34,24 +34,32 @@
     private transient AuthorConfiguration authorConfig;
     private transient boolean isStandaloneConfigIgnored;
     private transient List<CommitHash> ignoreCommitList;
+    private transient boolean isFormatsOverriding;
+    private transient boolean isIgnoreGlobListOverriding;
+    private transient boolean isIgnoreCommitListOverriding;
```
Comment (yamidark) [inline src/main/java/reposense/model/RepoConfiguration.java]: Let's add this check to the `equals` method as well

File: docs/UserGuide.md [hunk 8]
```diff
@@ -281,18 +281,21 @@ Repository's Location|Branch|File formats|Ignore Glob List|Ignore standalone con
 ---------------------|------|------------|----------------|------------------------|-------------------
 `https://github.com/foo/bar.git`|`master`|`java;css`|`test/**`|`yes`|`2fb6b9b2dd9fa40bf0f9815da2cb0ae8731436c7;c5a6dc774e22099cd9ddeb0faff1e75f9cf4f151`
 
+When using standalone config (it is not ignored), it is possible to override specific values from the standalone config by prepending the entered value with `override:`.
```
Comment (yamidark) [inline docs/UserGuide.md]: (**if** it is not ignored)

**Comment by myteo** [comment]
Changes made.

**Comment by myteo** [comment]
Requested changes made.

**Comment by eugenepeh** [comment]
Nits to the proposed message

```
In our implementation, Standalone config has the utmost priority; such
that the user is unable to override any of the fields without disabling
it.

However, there may be cases where user may want to import only certain
fields (e.g. authors) and discard/reconfigure the others. One example
of this can be to import the details of the authors, but change the
file formats of the analysis.

Let's provide the override functionality in the repo-config.csv file
using the prefix "override:".
```

**Pull request merged**